### PR TITLE
gapic-common changes for the diregapic, accompanied by a sample REST-only library

### PR DESCRIPTION
**Pull request by viacheslav-rostovtsev**

### CODE REVIEW

File: gapic-common/gapic-common.gemspec [hunk 1]
```diff
@@ -40,6 +40,7 @@ Gem::Specification.new do |spec|
   spec.add_dependency "googleapis-common-protos-types", ">= 1.0.4", "< 2.0"
   spec.add_dependency "googleauth", "~> 0.9"
   spec.add_dependency "grpc", "~> 1.25"
+  spec.add_dependency "faraday", "~> 1.3.0"
```
Comment (dazuma) [inline gapic-common/gapic-common.gemspec:43]: Are you sure you want to limit it to 1.3.x?

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        @endpoint = endpoint
+        @endpoint = "https://#{endpoint}" unless /https?:/.match? endpoint
+        @endpoint = @endpoint.chop if @endpoint.to_s.chars.last == "/"
+
+        @credentials = credentials
+      end
+
+      # @param uri [String]
+      # @return [Hash { Symbol => String }]
+      def make_get_request uri:, options:
```
Comment (dazuma) [inline gapic-common/lib/gapic/rest/client_stub.rb:38]: Do we intend for query parameters to be part of the URI string?

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        request_url = create_request_url uri
+        headers = {
+          "Authorization" => "Bearer #{@credentials.client.access_token}"
+        }
```
Comment (dazuma) [inline gapic-common/lib/gapic/rest/client_stub.rb:42]: We'll need a mechanism for some header cases, including the `x-goog-api-client` header.

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+        response = Faraday.get request_url do |req|
+          req.headers = headers
+        end
```
Comment (dazuma) [inline gapic-common/lib/gapic/rest/client_stub.rb:46]: Consider creating and reusing an actual Faraday connection object. That would allow configuring it (e.g. with timeout settings, etc.) at the ClientStub level, and also allow Faraday to use persistent connections if possible.

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+        result = response.to_hash
+        yield result, response.env if block_given?
+
+        if result[:status] != 200
+          raise ::Gapic::Rest::Error.new "An error has occurred when making a REST request", result: result
+        end
```
Comment (dazuma) [inline gapic-common/lib/gapic/rest/client_stub.rb:53]: Is there a way to avoid repeating all the request preparation and result handling in every method?

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        request_url = create_request_url uri
+        headers = {
+          "Authorization" => "Bearer #{@credentials.client.access_token}"
+        }
+
+        response = Faraday.get request_url do |req|
+          req.headers = headers
+        end
+
+        result = response.to_hash
```
Comment (dazuma) [inline gapic-common/lib/gapic/rest/client_stub.rb:48]: Why convert to a hash rather than just using the Faraday response object as-is?

**Review by dazuma — COMMENTED** [review]
A few initial thoughts on the gapic-common layer

File: gapic-common/gapic-common.gemspec (hunk 1 above)
Comment (viacheslav-rostovtsev) [reply gapic-common/gapic-common.gemspec]: fixed

File: gapic-common/lib/gapic/rest/client_stub.rb (hunk 2 above)
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/client_stub.rb]: updated

File: gapic-common/lib/gapic/rest/client_stub.rb (hunk 3 above)
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/client_stub.rb]: in progress

File: gapic-common/lib/gapic/rest/client_stub.rb (hunk 4 above)
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/client_stub.rb]: fixed

File: gapic-common/lib/gapic/rest/client_stub.rb (hunk 5 above)
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/client_stub.rb:41]: fixed

File: gapic-common/lib/gapic/rest/client_stub.rb (hunk 6 above)
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/client_stub.rb]: fixed

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/compute_small_services_pb.rb [hunk 7]
```diff
@@ -0,0 +1,76 @@
+# Generated by the protocol buffer compiler.  DO NOT EDIT!
```
Comment (dazuma) [inline shared/output/cloud/compute_small/lib/google/cloud/compute/v1/compute_small_services_pb.rb:1]: Can we not generate this file if we don't have a gRPC client?

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/service_stub.rb [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+module Google
+  module Cloud
+    module Compute
+      module V1
+        module Addresses
+          module Rest
+            ##
+            # REST service stub for the Addresses service.
+            #
+            class ServiceStub
```
Comment (dazuma) [inline shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/service_stub.rb:31]: General question: What goes in the `ServiceStub` class vs the `Client` class? Why not generate all the code directly into `Client` and not have this file at all?

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+              #   @param page_token [::String]
+              #     Specifies a page token to use. Set pageToken to the nextPageToken returned by a previous list request to get the next page of results.
+              #   @param project [::String]
+              #     Project ID for this request.
+              #   @param region [::String]
+              #     Name of the region for this request.
+              #
+              # @yield [result, env] Access the result along with the Faraday environment object
+              # @yieldparam result [::Google::Cloud::Compute::V1::AddressList]
+              # @yieldparam env [::Faraday::Env]
```
Comment (dazuma) [inline shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb:347]: I think you're actually yielding the Faraday::Response rather than the Faraday::Env. Can you double-check? And maybe determine which interface we want to expose? (I think the point of the block form is to provide access to the low-level response data such as response headers, for the uncommon use cases that care about it, so maybe decide based on completeness of the data that can be accessed?)

Also, can you document whether or not the block is called in the case of an error response? I think it should be called, but I'm not sure what the implementation currently does.

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+              #
+              # @raise [::Faraday::Error] if the REST call is aborted.
+              def list request, options = nil
+                raise ::ArgumentError, "request must be provided" if request.nil?
+
+                request = ::Gapic::Protobuf.coerce request, to: ::Google::Cloud::Compute::V1::ListAddressesRequest
+
+                @addresses_stub.list request, options: options do |response, env|
+                  yield response, env if block_given?
+                end
```
Comment (dazuma) [inline shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb:359]: What happens to successful calls that result in 4xx or 5xx responses? I think the gRPC side maps those to `Google::Cloud::Error` exceptions, but I'm not convinced the above code does the same.

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+              #   @param order_by [::String]
+              #     Sorts list results by a certain order. By default, results are returned in alphanumerical order based on the resource name.
+              #
+              #     You can also sort results in descending order based on the creation timestamp using `orderBy="creationTimestamp desc"`. This sorts results based on the `creationTimestamp` field in reverse chronological order (newest result first). Use this to sort resources like operations so that the newest operation is returned first.
+              #
+              #     Currently, only sorting by `name` or `creationTimestamp desc` is supported.
+              #   @param page_token [::String]
+              #     Specifies a page token to use. Set `pageToken` to the `nextPageToken` returned by a previous list request to get the next page of results.
```
Comment (dazuma) [inline shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb:178]: It looks like we don't have pagination helpers built into the client yet. Is that in scope?

What about long-running operations?

**Review by dazuma — CHANGES_REQUESTED** [review]

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+          conn.request :retry
+          conn.response :raise_error
+          conn.adapter :net_http
+        end
+
+        yield @connection if block_given?
+      end
+
+      # @param uri [String]
+      # @return [Hash { Symbol => String }]
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/client_stub.rb:47]: fixed

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+          conn.authorization :Bearer, @credentials.client.access_token
+          conn.request :retry
+          conn.response :raise_error
+          conn.adapter :net_http
+        end
+
+        yield @connection if block_given?
+      end
+
+      # @param uri [String]
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/client_stub.rb:46]: fixed

File: gapic-common/lib/gapic/rest/error_wrap.rb [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        # This parses the error message and the status code from the Google Cloud
+        # response's body and if successful, modifies the Faraday error to have
+        # :message and :status_code attributes, and then sets those attributes
+        # to the parsed values.
+        #
+        # This allows the Faraday Error to be correctly re-wrapped by ::Google::Cloud::Error
+        # and also surfaces the cloud error message instead of Faraday's generic one
+        #
+        def augment_faraday_error! err
+          return err unless err.response && err.response.is_a?(Hash)
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/error_wrap.rb:36]: noted! thanks for the tip

File: gapic-common/lib/gapic/rest/error_wrap.rb [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+          msg, status_code = try_parse_from_response err.response
+
+          unless msg.nil? && code.nil?
+            class << err
+              attr_accessor :message, :status_code
+            end
+
+            pre_msg = "An error has occurred when making a REST request: "
+            err.message = "#{pre_msg}#{msg}" unless msg.nil?
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/error_wrap.rb:45]: makes sense

File: gapic-common/lib/gapic/rest/error_wrap.rb [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+        end
+
+        private
+
+        # @param [Hash] response
+        def try_parse_from_response response
+          msg = nil
+          code = nil
+
+          code = response[:status] if response.key? :status
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/error_wrap.rb:60]: :+1:

File: gapic-common/lib/gapic/rest/error_wrap.rb [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+
+          if response.key?(:body) && response[:body]
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/error_wrap.rb:62]: :+1:

File: gapic-common/lib/gapic/rest/error_wrap.rb [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+          end
+
+          [msg, code]
+        end
+
+        ##
+        # Tries to get the error information from the JSON bodies
+        #
+        # @param [String] body_str
+        # @return [Array<String>]
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/error_wrap.rb:76]: the tuple type makes an unexpected (but welcome) appearance

File: gapic-common/lib/gapic/rest/error_wrap.rb [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+        def try_parse_from_body body_str
+          body = JSON.parse body_str
+          return [nil, nil] unless !body.nil? &&
+                                    body.key?("error") &&
+                                    body["error"].is_a?(Hash)
+
+          message = body["error"]["message"] if body["error"].key? "message"
+          code = body["error"]["message"] if body["error"].key? "code"
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/error_wrap.rb:84]: :+1:

File: gapic-common/lib/gapic/rest/error_wrap.rb [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+        ##
+        # Tries to get the error information from the JSON bodies
+        #
+        # @param [String] body_str
+        # @return [Array<String>]
+        def try_parse_from_body body_str
+          body = JSON.parse body_str
+          return [nil, nil] unless !body.nil? &&
+                                    body.key?("error") &&
+                                    body["error"].is_a?(Hash)
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/error_wrap.rb:81]: :+1:

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/compute_small_services_pb.rb (hunk 7 above)
Comment (viacheslav-rostovtsev) [reply shared/output/cloud/compute_small/lib/google/cloud/compute/v1/compute_small_services_pb.rb:1]: yes, I am going to update the Rake file with this

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/service_stub.rb (hunk 8 above)
Comment (viacheslav-rostovtsev) [reply shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/service_stub.rb:31]: This way clients for Rest and Grpc look symmetrical. 

On another level, 
`client` is an interface to the user, it deals with various input data,  and (for cloud) error shape (which is a way of saying 'various output data') and configuration
service_stub is an interface to the `gapic-common` rest layer, it deals with encoding and decoding data to/from json (and that data goes to the correct parts of http request, which is GRPC transcoding)
(then the client_stub deals with wrapping Faraday)

so it's cleaner this way, I think

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb (hunk 10 above)
Comment (viacheslav-rostovtsev) [reply shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb:359]: Faraday's error handling middleware wraps it into the Faraday::Error of the appropriate type and then it is going to be caught and wrapped to Google::Cloud::Error (for Cloud)

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb (hunk 11 above)
Comment (viacheslav-rostovtsev) [reply shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb:178]: No pagination, no LRO for the initial release. (LROs can be done manually: https://github.com/viacheslav-rostovtsev/google-showcase-dire/blob/fcb0acbbae98719de681865ef775040870a10800/test_use_showcase_lib/test/test_use_compute_alpha.rb#L35-L54)

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/service_stub.rb (hunk 8 above)
Comment (dazuma) [reply shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/service_stub.rb:31]: Symmetricality is good, but it's it's only really important in the interface. The `Service` and `Stub` classes on the gRPC side are generated by gRPC, and are "internal" for the purposes of GAPIC. They are not part of the actual client interface that a user should care about, and indeed we don't generate any documentation for them. Indeed, the only reason they exist at all is that gRPC is generating an extra interface layer that GAPIC has chosen to hide because we don't think it's good enough. We don't have that complexity on the REST side, and I don't see why we need to add it.

I could also see an argument for separating out `ServiceStub` if it were reasonable to expect its lower-level interface will be used by something other than the `Client` class. But I can't think of a use case at this point, and I'm inclined to argue YAGNI.

Comment (viacheslav-rostovtsev) [reply shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/service_stub.rb:31]: I would argue that it's still a cleaner separation of concerns and keeps the client class from trying to do too many things.

Comment (viacheslav-rostovtsev) [reply shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/service_stub.rb:31]: Also I disagree that the symmetry is important only for the interface. 
- it is for us when we try to reason about the code we generate as well. 
- Some (most?) end-users will end up reading the client class. It would be easier for them if they don't have to take in everything at once and if they have to read both clients, they would be able to go 'oh, this one is basically the same as the other one'.

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb (hunk 9 above)
Comment (viacheslav-rostovtsev) [reply shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb:347]: this whole yielding thing is in progress

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/service_stub.rb (hunk 8 above)
Comment (dazuma) [reply shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/service_stub.rb:31]: Let's discuss during the Ruby fixit time on Monday.

File: gapic-common/lib/gapic/rest/client_stub.rb [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+      def initialize endpoint:, credentials:
+        @endpoint = endpoint
+        @endpoint = "https://#{endpoint}" unless /https?:/.match? endpoint
+        @endpoint = @endpoint.chop if @endpoint.to_s.chars.last == "/"
+
+        @credentials = credentials
+
+        @connection = Faraday.new url: @endpoint do |conn|
+          conn.headers = { "Content-Type" => "application/json" }
+          conn.authorization :Bearer, @credentials.client.access_token
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/client_stub.rb:37]: should be fixed with the faraday middleware

File: gapic-common/lib/gapic/rest/error_wrap.rb [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+        # and also surfaces the cloud error message instead of Faraday's generic one
+        #
+        def augment_faraday_error! err
+          return err unless err.response && err.response.is_a?(Hash)
+          msg, status_code = try_parse_from_response err.response
+
+          unless msg.nil? && code.nil?
+            class << err
+              attr_accessor :message, :status_code
+            end
```
[the remark this answers is not in the document]
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/error_wrap.rb:42]: fixed by wrapping into an Gapic::Common::Rest::Error

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/service_stub.rb (hunk 8 above)
Comment (viacheslav-rostovtsev) [reply shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/service_stub.rb:31]: refactored into grpc_transcoding helper class

File: shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb (hunk 9 above)
Comment (viacheslav-rostovtsev) [reply shared/output/cloud/compute_small/lib/google/cloud/compute/v1/addresses/rest/client.rb:347]: yielding has been changed (the docs might be still stale?)

File: gapic-common/lib/gapic/rest/client_stub.rb (hunk 3 above)
Comment (viacheslav-rostovtsev) [reply gapic-common/lib/gapic/rest/client_stub.rb]: fixed
